flowbinding-common/src/main/java/reactivecircus/flowbinding/common/InitialValueFlow.kt

```diff
@@ -0,0 +1,30 @@
+package reactivecircus.flowbinding.common
+
+import androidx.annotation.RestrictTo
+import androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.onStart
+
+/**
+ * Converts a [Flow] to an [InitialValueFlow], taking an [initialValue] lambda for computing the initial value.
+ */
+@RestrictTo(LIBRARY_GROUP)
+@OptIn(ExperimentalCoroutinesApi::class)
+fun <T : Any> Flow<T>.asInitialValueFlow(initialValue: () -> T?): InitialValueFlow<T> = InitialValueFlow(
+    onStart {
+        initialValue()?.run { emit(this) }
```

Comment:
Does it make sense to support `null` as a valid initial value?

Reply:
I don't think so. This is a Library restricted function for converting a flow to an `InitialValueFlow` by passing in the `() -> T?` lambda. The return type of the lambda needs to be nullable because sometimes the initial value of a widget might be null e.g. `BottomNavigationViewItemSelectedFlow.kt`, in this case we skip the emission of the initial value instead of emitting null. Note that the `T` is bounded by `Any` which means `InitialValueFlow` can never emit `null`.

Reply:
Or do you think we should support `null` as the initial value?

Reply:
Yes, `null` should be a supported initial value. `null` emissions are supported by `Flow`, and there could be properties that are initially (or have simply been changed to) `null`.

Reply:
Which also means `T` should not be bounded to `Any`.

Reply:
I think there are only a handful of such bindings where a state can be unselected by setting it to `null`. Let me track them down and add docs & tests.

Reply:
@BenTilBrook @Svenjacobs what are some of the widgets with callback listeners where the params can be `null`? i.e. calling `widget.property = null` will trigger `onXxx(null)`?

I went through all the bindings and couldn't find one such binding 😕

Reply:
I don't know of any at the moment however maybe we should support `null` here for future compatibility? Since this is a generic function, specifying a type of `String` instead of `String?` for instance still ensures non-nullability of bindings that will never emit `null`.

Reply:
I think the following function signature should work for nullable as well as non-nullable types?

```kotlin
fun <T> Flow<T>.asInitialValueFlow(initialValue: () -> T): InitialValueFlow<T>
```

Reply:
Yeah we can remove the `: Any` bound, although restricting it to non-null now and making it nullable in the future if / when needed should not break source compatibility.

The `initialValue: () -> T?` needs to be nullable to support a couple of weird cases, e.g.

```
fun NavigationView.itemSelections(): InitialValueFlow<MenuItem> =
    callbackFlow {
        checkMainThread()
        val listener = NavigationView.OnNavigationItemSelectedListener { item ->
            safeOffer(item)
        }
        setNavigationItemSelectedListener(listener)
        awaitClose { setNavigationItemSelectedListener(null) }
    }
        .conflate()
        .asInitialValueFlow {
            var selectedItem: MenuItem? = null
            for (index in 0 until menu.size()) {
                val item = menu.getItem(index)
                if (item.isChecked) {
                    selectedItem = item
                    break
                }
            }
            selectedItem
        }
```

The `item` param of `OnNavigationItemSelectedListener` is always `NonNull`. But getting the `selectedItem` imperatively (e.g. in `asInitialValueFlow`) could return `null`, so in that case we can't emit anything as the initial value.

I don't thinking making it `InitialValueFlow<MenuItem?>` and emitting `null` as initial value if there's no selection makes sense here, as there's no `setNavigationItemUnselectedListener`. If the cases where an unselected listener is available, we usually expose custom sealed class type, which again means no nullable value.

Reply:
Interesting. So there are bindings that return an `InitialValueFlow` where however no initial value may exist? How does RxBinding handle this?

Reply:
This might be unexpected, or not? What if I call `skipInitialValue()` on such Flow. It would then skip the first value resulting from a user interaction.

Reply:
RxBinding does the [same thing](https://github.com/JakeWharton/RxBinding/blob/master/rxbinding-material/src/main/java/com/jakewharton/rxbinding4/material/NavigationViewItemSelectionsObservable.kt) 😄

Reply:
Wait RxBinding actually just emits a regular flow. But this can also be surprising as most of the other regular `Observable` bindings won't emit the current value initially?

Reply:
Maybe `InitialValueFlow` should only be returned where an initial value can be guaranteed? The behavior of other use cases like this specific case needs to be documented.

Reply:
There are only 3 such bindings:
- [BottomNavigationViewItemSelectionsObservable](https://github.com/JakeWharton/RxBinding/blob/master/rxbinding-material/src/main/java/com/jakewharton/rxbinding4/material/BottomNavigationViewItemSelectionsObservable.kt)
- [NavigationViewItemSelectionsObservable](https://github.com/JakeWharton/RxBinding/blob/master/rxbinding-material/src/main/java/com/jakewharton/rxbinding4/material/NavigationViewItemSelectionsObservable.kt)
- [TabLayoutSelectionEventObservable](https://github.com/JakeWharton/RxBinding/blob/master/rxbinding-material/src/main/java/com/jakewharton/rxbinding4/material/TabLayoutSelectionEventObservable.kt)

I don't mind making these regular `Flow` for consistency.

Reply:
[8607df4](https://github.com/ReactiveCircus/FlowBinding/pull/98/commits/8607df41cca5e7185040d0327f6707f83214c4d8)

Reply:
I'd prefer we emit `null` or something to indicate the absence of an initial selection. This makes the API more useful and consistent with all the other initial value-emitting APIs, as the rationale behind emitting the initial value by default is equally relevant.

Reply:
Basically by subscribing, you're saying "I want to subscribe to the state of this property". By not emitting initially under some circumstances, it creates a confusing API.

Reply:
@BenTilbrook how does emitting `null` initially make the API more useful? This would be inconsistent with both the widget's callback API which never emits `null`, and RxBinding which does not return `InitialValueObservable`.

Since there are only 3 of such bindings, I think it's fine to rely on documentation to highlight the optional initial value emission behavior?

Reply:
As discussed somewhat in #94, I think mirroring the underlying callback API is less useful than providing a reactive behavior adapter. Without the initial emission, in order to subscribe the state of selections, I need to invoke the imperative API AND the reactive API. It's not very ergonomic.

This is a bit of an edge case so I'm not going to push hard for it, but I think the above principles are good ones by which to design FlowBinding.

Reply:
I personally find it more confusing to have a binding emit nullable value only initially, as this behavior is not expressed in the type system and it forces the user to understand what `null` really means. I also don't find `null` to be a good representation of state as it can be interpreted differently in different contexts.

It's probably not a big deal either way as there are only 3 of them and I don't think we'll run into this issue with many of the new widgets being added to MDC.

Reply:
Now that you updated the three bindings that required a special null-handling, if you change the function signature to

```kotlin
fun <T : Any> Flow<T>.asInitialValueFlow(initialValue: () -> T): InitialValueFlow<T>
```

I'm fine with it :)

Reply:
Thanks @svenjacobs! It’s already updated😀